staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go

```diff
@@ -0,0 +1,366 @@
+		}
+		// only match /apis/<group>
+		if len(pathParts) == 2 {
+			r.groupDiscoveryHandler.ServeHTTP(w, req)
+			return
+		}
+
+		r.delegate.ServeHTTP(w, req)
+		return
+	}
+	if len(requestInfo.Subresource) > 0 {
+		http.NotFound(w, req)
+		return
+	}
+
+	customResourceName := requestInfo.Resource + "." + requestInfo.APIGroup
+	customResource, err := r.customResourceLister.Get(customResourceName)
+	if apierrors.IsNotFound(err) {
+		r.delegate.ServeHTTP(w, req)
+		return
+	}
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if customResource.Spec.Version != requestInfo.APIVersion {
+		r.delegate.ServeHTTP(w, req)
+		return
+	}
+	// TODO this is the point to do the condition checks
+
+	customResourceInfo := r.getServingInfoFor(customResource)
+	storage := customResourceInfo.storage
+	requestScope := customResourceInfo.requestScope
+	minRequestTimeout := 1 * time.Minute
+
+	switch requestInfo.Verb {
+	case "get":
+		handler := handlers.GetResource(storage, storage, requestScope)
+		handler(w, req)
+		return
+	case "list":
+		forceWatch := false
+		handler := handlers.ListResource(storage, storage, requestScope, forceWatch, minRequestTimeout)
+		handler(w, req)
+		return
+	case "watch":
+		forceWatch := true
+		handler := handlers.ListResource(storage, storage, requestScope, forceWatch, minRequestTimeout)
+		handler(w, req)
+		return
+	case "create":
+		handler := handlers.CreateResource(storage, requestScope, discovery.NewUnstructuredObjectTyper(nil), r.admission)
+		handler(w, req)
+		return
+	case "update":
+		handler := handlers.UpdateResource(storage, requestScope, discovery.NewUnstructuredObjectTyper(nil), r.admission)
+		handler(w, req)
+		return
+	case "patch":
+		handler := handlers.PatchResource(storage, requestScope, r.admission, unstructured.UnstructuredObjectConverter{})
+		handler(w, req)
+		return
+	case "delete":
+		allowsOptions := true
+		handler := handlers.DeleteResource(storage, allowsOptions, requestScope, r.admission)
+		handler(w, req)
+		return
+
+	default:
+		http.Error(w, fmt.Sprintf("unhandled verb %q", requestInfo.Verb), http.StatusMethodNotAllowed)
+		return
+	}
+}
+
+// removeDeadStorage removes REST storage that isn't being used
+func (r *customResourceHandler) removeDeadStorage() {
+	// these don't have to be live.  A snapshot is fine
+	// if we wrongly delete, that's ok.  The rest storage will be recreated on the next request
+	// if we wrongly miss one, that's ok.  We'll get it next time
+	storageMap := r.customStorage.Load().(map[types.UID]*customResourceInfo)
+	allCustomResources, err := r.customResourceLister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	for uid := range storageMap {
+		found := false
+		for _, customResource := range allCustomResources {
+			if customResource.UID == uid {
+				found = true
+				break
+			}
+		}
+		if !found {
+			delete(storageMap, uid)
+		}
+	}
+
+	r.storageMutationLock.Lock()
```

Comment:
> we read it above and then write here. Does it matter that we could race writing a new storage map?

I don't think so.  You get a few one offs of these created, but they ought to be GC-able, right?